Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/architecture/util.cpp

```diff
@@ -0,0 +1,134 @@
+#include "util.hpp"
+
+namespace sne
+{
+
+    bool operator==(const glm::vec3 &v1, const glm::vec3 &v2)
+    {
+        return v1[0] == v2[0] && v1[1] == v2[1] && v1[2] == v2[2];
+    }
+
+    bool operator<(const glm::vec3 &v1, const glm::vec3 &v2) // TO DISCUSS
+    {
+        // if ((*this) == v)
+        //     return false;
+
+        // double norm1 = norm(*this),
+        //        norm2 = norm(v);
+
+        // // TODO later:
+        // // Define with our own criteria (x pref, y pref and z pref?)
+        // return norm1 < norm2;
+
+        // Critera for delta:
+        return v1[0] < v2[0] && v1[1] < v2[1] && v1[2] < v2[2];
+    }
+
+    bool operator>(const glm::vec3 &v1, const glm::vec3 &v2)
+    {
+        return v2 < v1;
+    }
+
+    bool operator>=(const glm::vec3 &v1, const glm::vec3 &v2)
+    {
+        return v2 < v1 || v1 == v2;
+    }
+
+    bool operator<=(const glm::vec3 &v1, const glm::vec3 &v2)
+    {
+        return v2 > v1 || v2 == v1;
+    }
+
+    glm::vec3 operator-(const glm::vec3 &v1, const glm::vec3 &v2)
+    {
+        glm::vec3 res{
+            v1[0] - v2[0],
+            v1[1] - v2[1],
+            v1[2] - v2[2]};
+
+        return res;
+    }
+
+    glm::vec3 operator+(const glm::vec3 &v1, const glm::vec3 &v2)
+    {
+        glm::vec3 res{
+            v1[0] + v2[0],
+            v1[1] + v2[1],
+            v1[2] + v2[2]};
+
+        return res;
+    }
+
+    glm::vec3 operator+(const glm::vec3 &p, float &eps)
+    {
+        return glm::vec3{p[0] + eps, p[1] + eps, p[2] + eps};
+    }
+
+    glm::vec3 operator-(const glm::vec3 &p, float &eps)
+    {
+        return glm::vec3{p[0] - eps, p[1] - eps, p[2] - eps};
+    }
+
+    // glm::vec3 glm::vec3::operator*(double coeff) const
+    // {
+    //     glm::vec3 v = *this;
+    //     v[0] = v[0] * coeff;
+    //     v[1] = v[1] * coeff;
+    //     v[2] = v[2] * coeff;
+    //     return v;
+    // }
+
+    std::ostream &operator<<(std::ostream &oss, const glm::vec3 &v)
+	{
+		oss << "(" << v[0] << ", " << v[1] << ", " << v[2] << ")";
+		return oss;
+	}
+
+    glm::mat3 buildRotationMatrix(float alpha, float beta, float gamma)
```

Comment:
do you really need to use mat3 ? Is there an advantage of using separate matrix for rotation?

Reply:
To keep track:
As we discussed earlier, it avoid creating multiple times the matrix + the associate calcul, so it is more efficient and not really more complexe.